plugins/inputs/nfsclient/nfsclient.go

```diff
@@ -0,0 +1,421 @@
+	"CLOSE",
+	"SETATTR",
+	"FSINFO",
+	"RENEW",
+	"SETCLIENTID",
+	"SETCLIENTID_CONFIRM",
+	"LOCK",
+	"LOCKT",
+	"LOCKU",
+	"ACCESS",
+	"GETATTR",
+	"LOOKUP",
+	"LOOKUP_ROOT",
+	"REMOVE",
+	"RENAME",
+	"LINK",
+	"SYMLINK",
+	"CREATE",
+	"PATHCONF",
+	"STATFS",
+	"READLINK",
+	"READDIR",
+	"SERVER_CAPS",
+	"DELEGRETURN",
+	"GETACL",
+	"SETACL",
+	"FS_LOCATIONS",
+	"RELEASE_LOCKOWNER",
+	"SECINFO",
+	"FSID_PRESENT",
+	"EXCHANGE_ID",
+	"CREATE_SESSION",
+	"DESTROY_SESSION",
+	"SEQUENCE",
+	"GET_LEASE_TIME",
+	"RECLAIM_COMPLETE",
+	"LAYOUTGET",
+	"GETDEVICEINFO",
+	"LAYOUTCOMMIT",
+	"LAYOUTRETURN",
+	"SECINFO_NO_NAME",
+	"TEST_STATEID",
+	"FREE_STATEID",
+	"GETDEVICELIST",
+	"BIND_CONN_TO_SESSION",
+	"DESTROY_CLIENTID",
+	"SEEK",
+	"ALLOCATE",
+	"DEALLOCATE",
+	"LAYOUTSTATS",
+	"CLONE",
+}
+
+var nfsopFields = []string{
+	"ops",
+	"trans",
+	"timeouts",
+	"bytes_sent",
+	"bytes_recv",
+	"queue_time",
+	"response_time",
+	"total_time",
+	"errors",
+}
+
+func convert(line []string) []int64 {
+	var nline []int64
+	for _, l := range line[1:] {
+		f, _ := strconv.ParseInt(l, 10, 64)
+		nline = append(nline, f)
+	}
+	return nline
+}
+
+func in(list []string, val string) bool {
+	for _, v := range list {
+		if v == val {
+			return true
+		}
+	}
+	return false
+}
+
+func (n *NFSClient) parseStat(mountpoint string, export string, version string, line []string, fullstat bool, nfs3Ops map[string]bool, nfs4Ops map[string]bool, acc telegraf.Accumulator) error {
+	tags := map[string]string{"mountpoint": mountpoint, "serverexport": export}
+	nline := convert(line)
+	first := strings.Replace(line[0], ":", "", 1)
+
+	var fields = make(map[string]interface{})
+
+	if fullstat && first == "events" && len(nline) >= len(eventsFields) {
+		for i, t := range eventsFields {
+			fields[t] = nline[i]
+		}
+		acc.AddFields("nfs_events", fields, tags)
+	} else if fullstat && first == "bytes" && len(nline) >= len(bytesFields) {
+		for i, t := range bytesFields {
+			fields[t] = nline[i]
+		}
+		acc.AddFields("nfs_bytes", fields, tags)
+	} else if fullstat && first == "xprt" && len(line) > 1 {
+		switch line[1] {
+		case "tcp":
+			if len(nline)+2 >= len(xprttcpFields) {
+				for i, t := range xprttcpFields {
+					fields[t] = nline[i+2]
+				}
+				acc.AddFields("nfs_xprt_tcp", fields, tags)
+			}
+		case "udp":
+			if len(nline)+2 >= len(xprtudpFields) {
+				for i, t := range xprtudpFields {
+					fields[t] = nline[i+2]
+				}
+				acc.AddFields("nfs_xprt_udp", fields, tags)
+			}
+		}
+	} else if fullstat {
+		if version == "3" {
+			if nfs3Ops[first] && (len(nline) <= len(nfsopFields)) {
+				for i, t := range nline {
+					item := fmt.Sprintf("%s_%s", first, nfsopFields[i])
+					fields[item] = t
+				}
+				acc.AddFields("nfs_ops", fields, tags)
+			}
+		} else if version == "4" {
+			if nfs4Ops[first] && (len(nline) <= len(nfsopFields)) {
+				for i, t := range nline {
+					item := fmt.Sprintf("%s_%s", first, nfsopFields[i])
+					fields[item] = t
+				}
```

Comment:
Isn't this identical to the version 3 loop? Why do you need to distinguish both then? If it's the length-check, please do it beforehand and issue an error or similar.

Reply:
Old code again, I'll look at refactoring this.

Reply:
This one is still open.